### PR TITLE
Webserver: HTTPS support

### PR DESCRIPTION
**Pull request by wsnipex**
Enable HTTPS support in our webserver if available.

## Description

This is kept rather simple, so at least for now there is no gui setting to select certificate and key files.
Instead the user is required to create special://userdata/server.key and special://userdata/server.pem.
Cert creation could be handled in an addon.

## Motivation and Context

Security is always good

## How Has This Been Tested?

used the chorus webinterface via https on ubuntu linux.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Improvement (non-breaking change which improves existing functionality)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

suggestions for a better help text are very welcome ;)

### CODE REVIEW

**Comment by lrusak** [comment]
What about making this mandatory? Can we create certs if they don't exist? It would be nice to avoid adding another setting 😁

**Comment by Montellese** [comment]
I had this running years ago (see https://github.com/Montellese/xbmc/commits/webserver_https) but back then linux was an old version of microhttpd which didn't support HTTPS yet and it required a special build for win32. These days we can easily get this working on linux but AFAIK (at least a few months ago) the windows version doesn't support SSL/HTTPS.

If you want to support it on at least non-win32 platforms you should adjust the setting so that it is only available if microhttpd supports SSL/HTTPS.

**Comment by chewitt** [comment]
Making HTTPS mandatory requires self-signed certs to be created, which will result in all users seeing untrusted cert warning messages in their browser, which many will consider to be a step backwards for OOB useability. Services like letsencrypt that provide free trusted certs must be able to query the webserver through public FQDN during cert creation so they are not an option (and exposing Kodi to the internet should be discouraged). Creating self-signed certs also needs an option to set the hostname (cn) attribute and some basic cert management, i.e. regenerate after changing cn, which will add settings clutter. IMHO it's best to support this via as.xml as a first step until we learn what that issues are. GUI can always be added in a later release.

**Comment by wsnipex** [comment]
re windows: I would have hoped someone was willing to build microhttpd with gnutls support, depends platforms already do that

I really think kodi core should not try to create certificates in any way. If we want to make it easier for users, lets do it in an addon.

**Comment by Montellese** [comment]
@wsnipex back when I first looked into this I actually built a version of microhttpd with gnutls support using MinGW. Since then @Karlson2k has migrated this to a pure Visual Studio based build which didn't support any kind of SSL / HTTPS last time I checked. Integrating gnutls there will probably be very tricky and require quite some changes to microhttpd itself :-/

File: xbmc/network/WebServer.cpp [hunk 1]
```diff
@@ -1110,14 +1112,94 @@ static void logFromMHD(void* unused, const char* fmt, va_list ap)
...
+  XFILE::auto_buffer buf;
+  const char* keyFile = "special://userdata/server.key";
+  const char* certFile = "special://userdata/server.pem";
+
+  if (!file.Exists(keyFile) || !file.Exists(certFile))
+    return false;
+
+  if (file.LoadFile(keyFile, buf) > 0)
+  {
+    std::string keyPem(buf.get(), buf.length());
```
Comment (notspiff) [inline xbmc/network/WebServer.cpp]: skey.resize(buf.length());
skey.assign(buf.get());

File: xbmc/network/WebServer.cpp [hunk 2]
```diff
@@ -1110,14 +1112,94 @@ static void logFromMHD(void* unused, const char* fmt, va_list ap)
...
+    skey = keyPem;
+    file.Close();
+  }
+  else
+    CLog::Log(LOGDEBUG, "WebServer %s: Error loading: %s", __FUNCTION__, keyFile);
+
+  if (file.LoadFile(certFile, buf) > 0)
+  {
+    std::string certPem(buf.get(), buf.length());
```
Comment (notspiff) [inline xbmc/network/WebServer.cpp]: as above

File: xbmc/network/WebServer.cpp [hunk 3]
```diff
@@ -1110,14 +1112,94 @@ static void logFromMHD(void* unused, const char* fmt, va_list ap)
...
+    scert = certPem;
+    file.Close();
+  }
+  else
+    CLog::Log(LOGDEBUG, "WebServer %s: Error loading: %s", __FUNCTION__, certFile);
+
+  if (skey.size() > 0 && scert.size() > 0)
```
Comment (notspiff) [inline xbmc/network/WebServer.cpp]: !skey.empty() && !scert.empty()

File: xbmc/network/WebServer.cpp [hunk 4]
```diff
@@ -1110,14 +1112,94 @@ static void logFromMHD(void* unused, const char* fmt, va_list ap)
...
 struct MHD_Daemon* CWebServer::StartMHD(unsigned int flags, int port)
 {
   unsigned int timeout = 60 * 60 * 24;
+  const char* ciphers = "NORMAL:-VERS-TLS1.0";
 
 #if MHD_VERSION >= 0x00040500
   MHD_set_panic_func(&panicHandlerForMHD, nullptr);
 #endif
 
+  if (LoadCert(m_key, m_cert) && MHD_is_feature_supported(MHD_FEATURE_SSL) == MHD_YES)
```
Comment (notspiff) [inline xbmc/network/WebServer.cpp]: reverse if, no need to load cert if not supported.

File: xbmc/network/WebServer.cpp [hunk 5]
```diff
@@ -1110,14 +1112,94 @@ static void logFromMHD(void* unused, const char* fmt, va_list ap)
...
+#if (MHD_VERSION >= 0x00095207)
+                          | MHD_USE_INTERNAL_POLLING_THREAD /* MHD_USE_THREAD_PER_CONNECTION must be used only with MHD_USE_INTERNAL_POLLING_THREAD since 0.9.54 */
+#endif
+#if (MHD_VERSION >= 0x00040001)
+                          | MHD_USE_DEBUG /* Print MHD error messages to log */
+#endif
+                          | MHD_USE_SSL
+                          ,
+                          port,
+                          NULL,
```
Comment (notspiff) [inline xbmc/network/WebServer.cpp]: nullptr

**Review by notspiff — CHANGES_REQUESTED** [review]
looks good apart from the minors

**Comment by wsnipex** [comment]
thanks, fixed the minors

**Comment by notspiff** [comment]
afaict you missed the if reversal

**Comment by wsnipex** [comment]
imho it was already in the 2nd commit from the beginning. If that's not what you mean, then I don't get it

**Comment by notspiff** [comment]
ah yes. you are right, it is rewritten in second commit.

**Review by notspiff — APPROVED** [review]

**Comment by afedchin** [comment]
> re windows: I would have hoped someone was willing to build microhttpd with gnutls support, depends platforms already do that

why not use openssl? leave gnutls for GNU systems.

**Comment by Montellese** [comment]
> why not use openssl? leave gnutls for GNU systems.
because microhttpd only has built-in support for gnutls so you would have to write the openssl integration from scratch and someone manage to support both gnutls and openssl.

**Comment by afedchin** [comment]
I'll live without https then :)

**Comment by MartijnKaijser** [comment]
@karlson2k what's your take on this?

**Comment by wsnipex** [comment]
can someone please test that this doesn't break on windows? The option should be hidden.

**Review by Rechi — APPROVED** [review]
setting doesn't show up on Windows
code also looks good

**Comment by rmrector** [comment]
I have to suggest this not be added to the GUI. I agree that "exposing Kodi to the internet should be discouraged", but a setting in the GUI to enable SSL (secure communication) for the web server will encourage some users to do just that. Packed away in an advancedsetting, it could be accompanied with a wiki entry that clearly states it's not for external use, and can link to a couple of better options if remote access is needed.

I don't think it's all that useful on many networks that Kodi will be a part of, but if "not for external use" is made more clear then we are good. It should already be separated from a guest network or other public traffic if SMB/NFS, UPNP, or other mildly protected traffic is used.

**Comment by wsnipex** [comment]
simply enabling this setting is not enough, it requires some knowledge from the user to provide the cert.
And I strongly disagree that not having SSL support is better.

**Comment by rmrector** [comment]
I guess I'm concerned with the conflation of the SSL "security" provided when a standard user blindly follows the steps to create a cert and leave it unprotected in userdata (for this web server bolted on to an application that is meant for something else entirely) vs the real security that real web servers provide.

Encouraging standard users to click through the big red warnings that browsers present is bad. Not just for usability, but for their security on the internet everywhere, because they don't care about the details and will take that encouragement to other websites.

**Comment by wsnipex** [comment]
If you have any suggestion on how to improve then let's hear it, but I don't buy the "no security is better" argument.

**Comment by wsnipex** [comment]
jenkins build this please

**Pull request merged**